File: src/artefatos.c
// Trabalho realizado individualmente:
// Daniel Campos Soares - daniSoares08
//
// Nome do grupo no Canvas: RA3_10

/* Persistencia dos artefatos obrigatorios da ultima execucao. */

#include <stdio.h>
#include <stdlib.h>

#include "analisador_semantico.h"

static int criar_diretorio_resultados(void) {
#ifdef _WIN32
    return system("if not exist resultados mkdir resultados");
#else
    return system("mkdir -p resultados");
#endif
}

static int escrever_texto(const char *caminho, const char *texto) {
    FILE *arquivo = fopen(caminho, "w");

    if (arquivo == NULL) {
        return 1;
    }

    fputs(texto ? texto : "", arquivo);
    fclose(arquivo);
    return 0;
}

static int escrever_tabela(const ResultadoTabelaSimbolos *tabela_simbolos) {
    FILE *arquivo = fopen("resultados/tabela_simbolos.md", "w");
    size_t i;
    size_t j;

    if (arquivo == NULL) {
        return 1;
    }

    fprintf(arquivo, "# Tabela de simbolos\n\n");
    fprintf(arquivo, "| Identificador | Tipo | Escopo | Linha definicao | Linhas de uso |\n");
    fprintf(arquivo, "|---|---|---|---:|---|\n");
    for (i = 0; i < tabela_simbolos->quantidade_simbolos; i++) {
        const Simbolo *s = &tabela_simbolos->simbolos[i];
        fprintf(
            arquivo,
            "| %s | %s | %s | %d | ",
            s->identificador,
            nome_tipo_dado(s->tipo),
            s->escopo,
            s->linha_definicao
        );
        if (s->quantidade_usos == 0) {
            fprintf(arquivo, "-");
        } else {
            for (j = 0; j < s->quantidade_usos; j++) {
                fprintf(arquivo, "%s%d", j == 0 ? "" : ", ", s->linhas_uso[j]);
            }
        }
        fprintf(arquivo, " |\n");
    }
    fclose(arquivo);
    return 0;
}

static int escrever_erros(const ErroAnalise *erros, size_t quantidade_erros_semanticos) {
    FILE *arquivo = fopen("resultados/erros_semanticos.md", "w");
    size_t i;

    if (arquivo == NULL) {
        return 1;
    }

    fprintf(arquivo, "# Erros semanticos\n\n");
    if (quantidade_erros_semanticos == 0) {
        fprintf(arquivo, "Nenhum erro semantico registrado.\n");
    } else {
        for (i = 0; i < quantidade_erros_semanticos; i++) {
            fprintf(
                arquivo,
                "- linha %d, elemento `%s`: %s\n",
                erros[i].linha,
                erros[i].elemento,
                erros[i].causa
            );
        }
    }

    fclose(arquivo);
    return 0;
}

static void escrever_no(FILE *arquivo, const NoAst *no, int nivel) {
    int i;

    if (no == NULL) {
        return;
    }

    for (i = 0; i < nivel; i++) {
        fprintf(arquivo, "  ");
    }
    fprintf(
        arquivo,
        "- %s valor=`%s` operador=`%s` tipo=%s linha=%d\n",
        nome_tipo_no(no->tipo),
        no->valor,
        no->operador,
        nome_tipo_dado(no->tipo_dado),
        no->linha
    );

    escrever_no(arquivo, no->esquerda, nivel + 1);
    escrever_no(arquivo, no->direita, nivel + 1);
    escrever_no(arquivo, no->terceiro, nivel + 1);
    escrever_no(arquivo, no->proximo, nivel);
}

static int escrever_arvore(const ArvoreAtribuida *arvore_atribuida) {
    FILE *arquivo = fopen("resultados/arvore_sintatica_atribuida.md", "w");

    if (arquivo == NULL) {
        return 1;
    }

    fprintf(arquivo, "# Arvore sintatica atribuida\n\n");
    fprintf(arquivo, "%s\n\n", arvore_atribuida->descricao);
    escrever_no(arquivo, arvore_atribuida->raiz, 0);
    fclose(arquivo);
    return 0;
}

int salvarArtefatosUltimaExecucao(
    const char *arquivo_teste,
    const ResultadoTabelaSimbolos *tabela_simbolos,
    const ErroAnalise *erros_semanticos,
    size_t quantidade_erros_semanticos,
    const ArvoreAtribuida *arvore_atribuida,
    const ResultadoAssembly *assembly
) {
    /*
     * Salva os artefatos finais exigidos pelo enunciado.
     *
     * IMPLEMENTAR/COMPLETAR conforme as regras do trabalho:
     * - Tabela de simbolos da ultima execucao.
     * - Relatorio de erros semanticos, mesmo vazio.
     * - Arvore sintatica atribuida da ultima execucao.
     * - Assembly da ultima execucao semanticamente valida.
     * - Indicacao explicita do arquivo de teste usado nos artefatos finais.
     */
    FILE *arquivo;
    int falhou = 0;

    if (criar_diretorio_resultados() != 0) {
        return 1;
    }

    arquivo = fopen("resultados/arquivo_teste_final.md", "w");
    if (arquivo == NULL) {
        falhou = 1;
    } else {
        fprintf(arquivo, "# Arquivo de teste final\n\n%s\n", arquivo_teste);
        fclose(arquivo);
    }

    falhou |= escrever_tabela(tabela_simbolos);
    falhou |= escrever_erros(erros_semanticos, quantidade_erros_semanticos);
    falhou |= escrever_arvore(arvore_atribuida);
    falhou |= escrever_texto(
        "resultados/codigo_assembly_ultima_execucao.s",
        assembly && assembly->codigo
            ? assembly->codigo
            : "@ GERADO: nenhum Assembly disponivel nesta execucao.\n"
    );

    arquivo = fopen("resultados/relatorio_execucao.md", "w");
    if (arquivo == NULL) {
        falhou = 1;
    } else {
        fprintf(arquivo, "# Relatorio final de execucao\n\n");
        fprintf(arquivo, "- Arquivo analisado: `%s`\n", arquivo_teste);
        fprintf(arquivo, "- Simbolos registrados: %zu\n", tabela_simbolos->quantidade_simbolos);
        fprintf(arquivo, "- Erros registrados: %zu\n", quantidade_erros_semanticos);
        fclose(arquivo);
    }

    return falhou;
}

File: analisador/artefatos.py
# Trabalho realizado individualmente:
# Daniel Campos Soares - daniSoares08
#
# Nome do grupo no Canvas: RA3_10

"""Persistencia dos artefatos obrigatorios da ultima execucao."""

from __future__ import annotations

from pathlib import Path

from analisador.modelos import (
    ArvoreAtribuida,
    ErroAnalise,
    NoAst,
    ResultadoTabelaSimbolos,
    nome_tipo_dado,
    nome_tipo_no,
)


PROJETO_RAIZ = Path(__file__).resolve().parents[1]
RESULTADOS_DIR = PROJETO_RAIZ / "resultados"


def _escrever_texto(caminho: Path, texto: str) -> int:
    try:
        caminho.write_text(texto or "", encoding="utf-8")
    except OSError:
        return 1
    return 0


def _escrever_tabela(tabela_simbolos: ResultadoTabelaSimbolos) -> int:
    linhas = [
        "# Tabela de simbolos\n\n",
        "| Identificador | Tipo | Escopo | Linha definicao | Linhas de uso |\n",
        "|---|---|---|---:|---|\n",
    ]
    for simbolo in tabela_simbolos.simbolos:
        usos = "-" if not simbolo.linhas_uso else ", ".join(str(uso) for uso in simbolo.linhas_uso)
        linhas.append(
            f"| {simbolo.identificador} | {nome_tipo_dado(simbolo.tipo)} | "
            f"{simbolo.escopo} | {simbolo.linha_definicao} | {usos} |\n"
        )
    return _escrever_texto(RESULTADOS_DIR / "tabela_simbolos.md", "".join(linhas))


def _escrever_erros(erros: list[ErroAnalise]) -> int:
    linhas = ["# Erros semanticos\n\n"]
    if not erros:
        linhas.append("Nenhum erro semantico registrado.\n")
    else:
        for erro in erros:
            linhas.append(
                f"- linha {erro.linha}, elemento `{erro.elemento}`: {erro.causa}\n"
            )
    return _escrever_texto(RESULTADOS_DIR / "erros_semanticos.md", "".join(linhas))


def _escrever_no(no: NoAst | None, nivel: int, linhas: list[str]) -> None:
    if no is None:
        return
    linhas.append(
        f"{'  ' * nivel}- {nome_tipo_no(no.tipo)} valor=`{no.valor}` "
        f"operador=`{no.operador}` tipo={nome_tipo_dado(no.tipo_dado)} "
        f"linha={no.linha}\n"
    )
    _escrever_no(no.esquerda, nivel + 1, linhas)
    _escrever_no(no.direita, nivel + 1, linhas)
    _escrever_no(no.terceiro, nivel + 1, linhas)
    _escrever_no(no.proximo, nivel, linhas)


def _escrever_arvore(arvore_atribuida: ArvoreAtribuida) -> int:
    linhas = [
        "# Arvore sintatica atribuida\n\n",
        f"{arvore_atribuida.descricao}\n\n",
    ]
    _escrever_no(arvore_atribuida.raiz, 0, linhas)
    return _escrever_texto(
        RESULTADOS_DIR / "arvore_sintatica_atribuida.md",
        "".join(linhas),
    )


def salvarArtefatosUltimaExecucao(
    arquivo_teste,
    tabela_simbolos: ResultadoTabelaSimbolos,
    erros_semanticos: list[ErroAnalise],
    arvore_atribuida: ArvoreAtribuida,
    assembly,
):
    """Salva os artefatos finais exigidos pelo enunciado."""
    falhou = 0
    try:
        RESULTADOS_DIR.mkdir(exist_ok=True)
    except OSError:
        return 1

    falhou |= _escrever_texto(
        RESULTADOS_DIR / "arquivo_teste_final.md",
        f"# Arquivo de teste final\n\n{arquivo_teste}\n",
    )
    falhou |= _escrever_tabela(tabela_simbolos)
    falhou |= _escrever_erros(erros_semanticos)
    falhou |= _escrever_arvore(arvore_atribuida)
    falhou |= _escrever_texto(
        RESULTADOS_DIR / "codigo_assembly_ultima_execucao.s",
        assembly or "@ GERADO: nenhum Assembly disponivel nesta execucao.\n",
    )
    falhou |= _escrever_texto(
        RESULTADOS_DIR / "relatorio_execucao.md",
        "# Relatorio final de execucao\n\n"
        f"- Arquivo analisado: `{arquivo_teste}`\n"
        f"- Simbolos registrados: {len(tabela_simbolos.simbolos)}\n"
        f"- Erros registrados: {len(erros_semanticos)}\n",
    )
    return falhou
